### PR TITLE
Multiple scopes in Arguments and Bind Scope

### PR DESCRIPTION
**Pull request by proux01**
The `Arguments` command only supported at most one scope for each argument. This PR enables multiple scopes. This would be useful for instance in MathComp where we'd like to bind types like `int` or `rat` to both the generic `ring_scope` and the specifics `int_scope` or `rat_scope` (for instance operators like `+` or `*` should be the generic ring ones in both cases but parsing of constants can be specific to int or rat). The change should be fully backward compatible.

- [x] Added / updated **test-suite**.

- [x] Added **changelog**.
- [x] Added / updated **documentation**.
  
  - [ ] ~Documented any new / changed **user messages**.~
  - [x] Updated **documented syntax** by running `make -f Makefile.dune doc_gram_rsts`.

- [x] Opened **overlay** pull requests.

#### Overlays (to be merged in sync with the PR)
* https://github.com/LPCIC/coq-elpi/pull/389

### CODE REVIEW

**Comment by gares** [comment]
From the doc I don't get which scope is at the top of the stack (as precedence) when `x` is interpreted when I say:
```
Arguments f x%A%B
```
Similarly `add_inside` and `add_outside` have the same ambiguity (but I did not re-read that doc section).

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (jfehrle) [inline doc/sphinx/user-extensions/syntax-extensions.rst:1456]: You need a sentence or two and perhaps an example to make "inside" and "outside" clear.  What does it mean to have multiple bindings and how would they be used?  Is it a set of bindings, a list of bindings, or something else?  It's not a stack, is it?  "inside" doesn't fit either of those.  Perhaps there's a different word that's clearer.

Optional: BTW, the existing example is wordy, the function seems too complex (distracting) and not very clear.  The key parts should be highlighted with in-line comments `(* ... *)`.  There should be an explanation of the effect on the Check output and maybe it should show what happens without the `Bind Scope`.  It would also be great to briefly explain the benefit/why users will want to use this command.  Are you willing to improve this?

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (jfehrle) [inline doc/sphinx/language/extensions/arguments-command.rst:88]: Please split this line

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (jfehrle) [inline doc/sphinx/language/extensions/arguments-command.rst:75]: These text descriptions need revision to reflect that there may be multiple scopes. is just `@name__1%scope__1%scope__2 ...`  Does "add_inside" just mean "before" (or "after")?

**Review by jfehrle — COMMENTED** [review]
Documentation suggestions

**Comment by proux01** [comment]
Thanks for the reviews, the doc was indeed insufficient and the naming of the attributes was bad. This should be better now.

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (jfehrle) [inline doc/sphinx/user-extensions/syntax-extensions.rst:1518]:
```suggestion
         ``bool`` and ``nat`` types, while the scope ``F_scope`` could
         override some of these notations specifically for
         ``bool`` and another ``F'_scope`` could override them
         specifically for ``nat``, which would then be bound to
         ``_%F'_scope%T_scope``.
```

Is the first _ in `_%F'_scope%T_scope` correct?

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (jfehrle) [inline doc/sphinx/user-extensions/syntax-extensions.rst:1461]: Great example, very clear!

**Review by jfehrle — COMMENTED** [review]
Much better on the doc.  A few more suggestions.

**Comment by gares** [comment]
Much better. I think we still need a review from syntax maintainers

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (proux01) [reply doc/sphinx/language/extensions/arguments-command.rst]: Sorry, done

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (proux01) [reply doc/sphinx/language/extensions/arguments-command.rst]: You're right, this is unrelated and doesn't bring anything here.

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (proux01) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: Good catch!

**Comment by proux01** [comment]
> I think we still need a review from syntax maintainers

That means @herbelin , @mattam82 or @ejgallego apparently.

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (jfehrle) [inline doc/sphinx/language/extensions/arguments-command.rst:55]: Would you put the entries back in their original order?  The entries are ordered to correspond to the order of the grammar productions.  Generally reordering should be done only if there's a good reason to do so--it makes it much harder to see what's been changed.  Looks like the original followed this principle imperfectly :-(.

Optional: `argument_spec` should be after `arg_specs` in the grammar

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (jfehrle) [inline doc/sphinx/language/extensions/arguments-command.rst:75]:
```suggestion
         :n:`(@name__1 @name__2 ...){* % @scope_key }` is shorthand for
         :n:`@name__1{* % @scope_key } @name__2%{* % @scope_key } ...`
```

A bit more readable.  Can you also do this for the other 2?  And add a new item to the list `:n:`{* %scope_key }` just before this item explaining how multiple scopes are interpreted.  Similar to the wording "When multiple scopes are attached to a type, notations are interpreted in the first scope containing them, from the top of the stack." used below.  Maybe just that sentence.

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (jfehrle) [inline doc/sphinx/language/extensions/arguments-command.rst:77]: I notice the grammar permits things like `( name1%scope1 name2 )%scope2`, which seems almost too flexible.  I suppose name1 gets `%scope1%scope2`.

**Review by jfehrle — COMMENTED** [review]
a couple more things

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (proux01) [reply doc/sphinx/language/extensions/arguments-command.rst:55]: Sorry, I forgot to explain that change. My goal was to make the order more consistent with the itemize just below the grammar. This indeed make things inconsistent with the grammar. I chose to fix that by reordering the grammar so that everything (grammar, itemize and precise description) is more consistent. Does that seem reasonable to you?

> [reordering] makes it much harder to see what's been changed

I agree, I tried to mitigate this by isolating the reordering in a dedicated commit.

> Optional: argument_spec should be after arg_specs in the grammar

Done

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (proux01) [reply doc/sphinx/language/extensions/arguments-command.rst]: Thanks, that's indeed much better.

> And add a new item [...]

Indeed, done

I also took the opportunity to fix the scope/scope_key glitch

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (proux01) [reply doc/sphinx/language/extensions/arguments-command.rst:78]: Good catch. This was obviously forbidden before and still is (gives `Error: scope declared twice`). I could either:
* adapt the grammar / document the error
* adapt the code to match the current doc / your interpretation

I'd lean toward the second solution but I'd appreciate your opinion (if any). In any case, this is just a matter of convenience, none of the options restricts the expressivity of the command.

Comment (jfehrle) [reply doc/sphinx/language/extensions/arguments-command.rst:78]: I suggest addressing that in another PR in which it can get a proper discussion.  Since it's not backward-compatible, the old form may need to be deprecated before it can be changed.

File: doc/tools/docgram/orderedGrammar
Comment (jfehrle) [inline doc/tools/docgram/orderedGrammar:799]: Please keep these lines after `"&"` as they were, also use that order in the description of the Arguments command.  Let's not change it without a reason.

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (jfehrle) [inline doc/sphinx/language/extensions/arguments-command.rst:66]: My mistake--this should have stayed where it was, but do include wording you added

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (jfehrle) [reply doc/sphinx/language/extensions/arguments-command.rst:78]: Though if it always gives an error now, then deprecation isn't needed.  But still better in a separate PR.

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (jfehrle) [reply doc/sphinx/language/extensions/arguments-command.rst]: > I also took the opportunity to fix the scope/scope_key glitch

I think that should be in a separate PR that explains what the "glitch" is and allow for proper discussion.  If scope names and scope keys can be used interchangeably almost everywhere, we should update the grammar/doc for all those uses consistently.

**Review by jfehrle — COMMENTED** [review]
not done yet

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (jfehrle) [reply doc/sphinx/language/extensions/arguments-command.rst:78]: Though if it always gives an error, deprecation is not needed.

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (proux01) [reply doc/sphinx/language/extensions/arguments-command.rst]: You are right, here it is: https://github.com/coq/coq/pull/16518

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (proux01) [reply doc/sphinx/language/extensions/arguments-command.rst:78]: Sorry, I kinda forgot that PR.

So, both options are perfectly backward compatible: the first one doesn't change the code and the second one is a strict extension of the current syntax (only more things are accepted). I rather think this should be done in the current PR as this is straightforward and not doing it would mean that this PR introduces a bug left to be fixed in a subsequent PR. So done, chosing the second option (i.e., adapt the code to match the current doc, c.f. last commit https://github.com/coq/coq/pull/16472/commits/9cebf31857ff844873a23e2b04e1344ea71a3345)

File: doc/tools/docgram/orderedGrammar
Comment (proux01) [reply doc/tools/docgram/orderedGrammar]: Well I explained my rationale above: https://github.com/coq/coq/pull/16472#discussion_r969319803 you can disagree with it, we can discuss it but ignoring it feels a bit rude. As I said, there is the following list just below the grammar of the `Arguments` command:

> * Making arguments implicit. Afterward, implicit arguments
>     must be omitted in any expression that applies :token:`reference`.
> * Declaring that some arguments of a given function should
>     be interpreted in a given scope.
> * Affecting when the :tacn:`simpl` and :tacn:`cbn` tactics unfold the function.
>     See :ref:`Args_effect_on_unfolding`.
> * Providing bidirectionality hints.  See :ref:`bidirectionality_hints`.

The first item correspond to all the `@name`, `(...)`, `[...]`and `{...}`, the second one to the `%scope` things, the third one to `/` and `!` while the last one corresponds to `&`. Hence the order I'd like to follow (but maybe we could do better).

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (jfehrle) [reply doc/sphinx/language/extensions/arguments-command.rst:78]: It seems to me that getting one of the syntax maintainers to review this point and the PR is should be a priority--their comments might impact the documentation, which is my main focus.

File: doc/tools/docgram/orderedGrammar
Comment (jfehrle) [reply doc/tools/docgram/orderedGrammar]: I think you initially missed a few of my earlier comments.  Should I call that rude?  I think not.  It's very easy to miss details, plus it's so easy to misinterpret electronic communications.  It's better just to raise a point again.  Your comments here have more detail that you didn't give earlier.

The following changes are inconsistent with what we've done elsewhere in the manual:
- Text describing grammar items should appear in exactly the order they appear in the grammar so it's easy to find.  So the text for ``` :n:`@name {? % @scope }` ``` is in the wrong place.  The grammar is sorted by dependencies, so `argument_spec` should be after `arg_specs`.
- The entry for `!` is in the wrong place compared to the grammar
- It's not essential that the bullet items you quote are in exactly the order matching the grammar.  The correspondence between the bullet items and the syntax descriptions is neither explicit in the text nor obvious (at least not to an expert).

The text doesn't say that `@name`, `(...)` is about implicit parameters at all.  How does it work?

Also I'd note that the description of the command is pretty weak; it doesn't tell you how to use the command.  And it is a bit of a hodge podge.  (That's not a request to improve it in this PR!)

We can consider whether to move `/` and `&` as you have.

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (jfehrle) [inline doc/sphinx/language/extensions/arguments-command.rst:76]: I notice the original text has `scope_key` in the grammar but uses `scope` in the descriptions.  Which one is correct? If you're willing, another thing to clean up.

File: test-suite/output/ArgumentsScope.v
Comment (proux01) [reply test-suite/output/ArgumentsScope.v:46]: correct, as is already the case with scope indications on any piece of code : `expr%A%B` means `(expr %A) %B`and as mentionned in the doc:
> When multiple scopes are present, notations are interpreted in the leftmost scope containing them.

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (proux01) [reply doc/sphinx/language/extensions/arguments-command.rst:76]: You're right. `scope` (that is either `scope_key` or `scope_name`) is the correct one according to that sentence:

> :token:`scope` can be either a scope name or its delimiting key

(and the code indeed accepts both), the discrepancy in the grammar is fixed by https://github.com/coq/coq/pull/16518

File: doc/tools/docgram/orderedGrammar
Comment (proux01) [reply doc/tools/docgram/orderedGrammar]: Thanks for the explanation. So indeed you are right that I should revert to the original order (except for `argument_spec` that should be after `arg_specs` in the grammar). What about the following:

1. reverting everything to original order
1. except for `argument_spec` that should come just after `arg_specs`
1. put the two first items of the bullet list last, to better match the order of the grammar (although this order doesn't have to match, I still feel it easier to read when it is more consistent)
2. this two first bullet should link to the parts of the manual about implicit arguments and scopes respectively

Regarding the entry for `!` I understand its position according to the grammar but being semantically much closer from `/`, it feels a bit strange when reading it. Shouldn't that warrant an exception to the rule? (`/` and `!` are syntactically in different places in the grammar because the first must appear only one while the last can be attached to any argument but they are both about the very same thing in practice) 

> The text doesn't say that @name, (...) is about implicit parameters at all. How does it work?

Indeed, this is only specified in the ``:n:`name`` entry 

> The construct :n:`@name {? % @scope }` declares :n:`@name` as non-implicit if `clear implicits` is spec\
ified or at least one other name is declared implicit in the same list of :n:`@name`\s.

Maybe this sentence should be copied or moved to the ``:n:`( ... )`` entry?

> Also I'd note that the description of the command is pretty weak; it doesn't tell you how to use the command. And it is a bit of a hodge podge. (That's not a request to improve it in this PR!)

There are examples in the subsections below, but maybe they need improvement (in a further PR indeed)

> We can consider whether to move / and & as you have.

Well, now that I understand the rules, I feel the original order is just as good.

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (herbelin) [reply doc/sphinx/language/extensions/arguments-command.rst:78]: Is there a specific question you want feedback on or a global review?

Comment (proux01) [reply doc/sphinx/language/extensions/arguments-command.rst:78]: More a global review / a confirmation that the new feature (extending `Arguments` to support multiple scopes) doesn't look stupid / fundamentally flawed.

File: doc/tools/docgram/orderedGrammar
Comment (jfehrle) [reply doc/tools/docgram/orderedGrammar]: I agree on 1 and 2 above.

>> We can consider whether to move / and & as you have.

> Well, now that I understand the rules, I feel the original order is just as good.

"moving" meaning changing the order both in the grammar and the text.  That would make them adjacent to `!`, which seems in line with your thinking.

> 3 put the two first items of the bullet list last

I don't see the benefit.  The bullet list is meant to be in descending order of importance/frequency of use.  IINM the order already follows that principle.

> 4 this two first bullet should link to the parts of the manual about implicit arguments and scopes respectively

In the first, maybe change "implicit arguments" to ```:term:`implicit arguments` ```.  (But note the implicit arguments chapter is not easy to read, either.  It needs an initial section introducing the main concepts with several short examples.  I think examples are usually easier for readers to grasp and the text is more for elaboration.  Hence the popularity of cut and paste.)

In the second, create a new glossary term by using ```:gdef:`scope` ``` at the target point of the link and then ```:term:`scope` ``` to refer to it.

> Maybe this sentence should be copied or moved to the ``:n:`( ... )`` entry?

Eh.  The sentence you refer to is hardly a model of clarity or completeness.  There should be a paragraph or two here with short examples explaining the "implicit" concept separate from the text describing the syntax .  A better, more succinct sentence could be repeated in multiple entries; in a reference manual, you want to minimize the amount of text that has to be read to understand something.  OTOH, we don't want to needlessly repeat somewhat complex points.  I would leave it alone for now.

Comment (proux01) [reply doc/tools/docgram/orderedGrammar:801]: > I agree on 1 and 2 above.

Done
I also took the opportunity to do a squash and rebase to clean the log.
 
> > > We can consider whether to move / and & as you have.
> 
> > Well, now that I understand the rules, I feel the original order is just as good.
> 
> "moving" meaning changing the order both in the grammar and the text. That would make them adjacent to `!`, which seems in line with your thinking.

Right, but then the `name` entry gets separated from the `( ... )`, `[ ... ]` and `{ ... }` ones, so that may not be really better. Let's stick with  the original order.
 
> > 3 put the two first items of the bullet list last
> 
> I don't see the benefit. The bullet list is meant to be in descending order of importance/frequency of use. IINM the order already follows that principle.

Ok, you're right.

> > 4 this two first bullet should link to the parts of the manual about implicit arguments and scopes respectively
> 
> In the first, maybe change "implicit arguments" to `` :term:`implicit arguments` ``.

Done (changing plural to singular as it's the way the term was defined)

> (But note the implicit arguments chapter is not easy to read, either. It needs an initial section introducing the main concepts with several short examples. I think examples are usually easier for readers to grasp and the text is more for elaboration. Hence the popularity of cut and paste.)

Ouch, indeed that section would need some major work to make it more accessible. It directly dives into technicality of how automatic inference of implicit arguments work without really explaining what implicit arguments are in the first place.
 
> In the second, you could create a new glossary term by using `` :gdef:`scope` `` at the target point of the link and then `` :term:`scope` `` to refer to it.

Thanks, done
 
> > Maybe this sentence should be copied or moved to the `` :n:`( ... ) `` entry?
> 
> Eh. The sentence you refer to is hardly a model of clarity or completeness. There should be a paragraph or two here with short examples explaining the "implicit" concept separate from the text describing the syntax . A better, more succinct sentence could be repeated in multiple entries; in a reference manual, you want to minimize the amount of text that has to be read to understand something. OTOH, we don't want to needlessly repeat somewhat complex points. I would leave it alone for now.

Ok, let's defer that to a further PR (the one on implicit arguments might be more urgent anyway).

Comment (jfehrle) [reply doc/tools/docgram/orderedGrammar:801]: > > In the first, maybe change "implicit arguments" to `` :term:`implicit arguments` ``.

> Done (changing plural to singular as it's the way the term was defined)

```:term:`implicit arguments <implicit argument>` ``` will show the plural with the correct link

**Review by jfehrle — APPROVED** [review]
LGTM.  Let's declare victory on the documentation

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (herbelin) [reply doc/sphinx/language/extensions/arguments-command.rst:72]: Since we can do `t%sc%sc'` in terms, supporting an implicit `%sc%sc'` looks indeed totally relevant.

In passing, there is the inconsistency that `%sc` in terms opens the scope for all subterms while `%sc` in `Arguments` opens the scope only for the immediate subterms. Ideally, both should be allowed in both terms and `Arguments` but this requires a syntax change. I don't know if this PR can be an opportunity to do something? (Possible notations for distinguishing the two that were discussed were `%%` for immediate scope and `%` for persistent opening, or `%` and `%!`; can a script be used to automatically change all `Arguments` from using `%` into say `%%`, or am I dreaming too far?).

Back to this PR, it thus looks good to me.

File: doc/tools/docgram/orderedGrammar
Comment (proux01) [reply doc/tools/docgram/orderedGrammar:801]: Indeed, thanks

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (proux01) [reply doc/sphinx/language/extensions/arguments-command.rst:72]: > In passing, there is the inconsistency that %sc in terms opens the scope for all subterms while %sc in Arguments opens the scope only for the immediate subterms.

I remember that discussion: https://github.com/coq/coq/pull/9123
I agree we should fix that discrepancy, but I think this belong to a further pull request.

**Comment by proux01** [comment]
@jfehrle thanks, would you be the assignee? Now that we got approval from extensible syntax maintainers https://github.com/coq/coq/pull/16472#discussion_r979240931 it seems we've reached a state where this can be merged.

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
merging soon

**Comment by Alizter** [comment]
@coqbot merge now

**Pull request merged**
